### DISCUSSION TITLE
Verified MFA doesn't set AAL to `aal2`

### DISCUSSION THREAD

**Discussion by hkamran80** [supabase/supabase#11383] [Questions] [answered]
In my Next.js app, I have a user flow that enrolls and verifies a user's MFA. When I sign in, the AAL is set to `aal1` for both current and next, which prevents requesting a 2FA token. It also prevents the user from unenrolling their verified MFA methods because of the required `aal2` level.

This is the JSON object I get when I call `supabase.auth.mfa.getAuthenticatorAssuranceLevel`:
```json
{
    "currentLevel": "aal1",
    "nextLevel": "aal1",
    "currentAuthenticationMethods": [
        {
            "method": "password",
            "timestamp": 1672514024
        }
    ]
}
```

And this is the output of `supabase.auth.mfa.listFactors`:
```json
{
    "all": [
        {
            "id": "460ee0bc-1f3e-xxxx-a128-01d6a07e140b",
            "created_at": "2022-12-31T15:12:56.170472Z",
            "updated_at": "2022-12-31T15:13:26.311767Z",
            "status": "verified",
            "factor_type": "totp"
        }
    ],
    "totp": [
        {
            "id": "460ee0bc-1f3e-xxxx-a128-01d6a07e140b",
            "created_at": "2022-12-31T15:12:56.170472Z",
            "updated_at": "2022-12-31T15:13:26.311767Z",
            "status": "verified",
            "factor_type": "totp"
        }
    ]
}
```

Does anybody have any idea as to why the AAL isn't being changed?

**Comment by tjmills-dev** [comment]
I've also experienced something similar - enrolling + validating while signed in results in aal2 being set correctly for both, but signing out then signing back in sets it to aal1 for both

**Reply by emileond** [reply]
Same here, I `getAuthenticatorAssuranceLevel` returns aal1 for both current and next levels. Hopefully this can get fixed soon. 🙁

**Reply by tjmills-dev** [reply]
@emileond I made an issue here https://github.com/supabase/gotrue-js/issues/589 - do leave a comment, might help diagnose the issue

**Reply by J0** [reply]
Hey all,

Auth team here - apologies for the inconvenience and confusion caused. Just want to let you know that we're looking into the issue right now and will get back to you as soon as we can

Jo

**Reply by tjmills-dev** [reply]
> Hey all,
> 
> Auth team here - apologies for the inconvenience and confusion caused. Just want to let you know that we're looking into the issue right now and will get back to you as soon as we can
> 
> Jo

Thanks!

**Reply by hkamran80** [reply]
> Hey all,
> 
> Auth team here - apologies for the inconvenience and confusion caused. Just want to let you know that we're looking into the issue right now and will get back to you as soon as we can
> 
> Jo

Thank you!

**Reply by J0** [reply]
Hey all! 

Thanks for your patience so far - I'm trying to reproduce this on NextJS 13 via the following 

1. `auth.signUp`
2. `.challenge` and then call `.verify()`
3. `auth.signOut`
4. `auth.signIn`
5. `auth.getAuthenticatorAssuranceLevel`

But having a little difficulty here - do y'all mind sharing your auth settings or anything else that might be worth noting? For reference here's what the auth options looks like. I'll try again with our helper `challengeAndVerify` method later as well to make sure there's nothing off there. 

Here's the auth settings I'm using - I'm on Chrome
```
const options = {
  db: {
    schema: 'public',
  },
  auth: {
    autoRefreshToken: true,
    persistSession: true,
    detectSessionInUrl: true
  }
```

**Reply by tjmills-dev** [reply]
@J0
Thanks for looking into it, my auth settings are just the default ones (which I think are the same as above), I'm on Firefox using Sveltekit
I've been testing it using challengeAndVerify, when I'll get chance later I'll try .challenge then .verify

**Reply by J0** [reply]
Thanks for the swift response! I’ll also download Firefox on my end and try

**Reply by tjmills-dev** [reply]
@J0 Just tried it on Chrome with separating the two methods - no dice :(

**Reply by J0** [reply]
Ok! thanks for the update

**Reply by emileond** [reply]
@J0 Thanks for taking a look at this. I'm also using the default auth configuration on Nextjs 12.2.5 - One odd thing I've noticed on Supabase is that when I try to enable Point in time recovery (PITR), it shows a message that says my project is too old to be able to enable PITR. So I wonder if something related to the project might be affecting MFA too:

<img width="442" alt="Screen Shot 2023-02-13 at 20 54 35" src="https://user-images.githubusercontent.com/29111636/218627027-cb4568fc-0f91-48b1-a462-9f4d83ceefcf.png">

My code looks almost exactly the same as the one [in this guide](https://supabase.com/docs/guides/auth/auth-mfa) - I'm using `challengeAndVerify` when enrolling, but the problem seems to be with `getAuthenticatorAssuranceLevel`.

Hope this helps.

**Reply by J0** [reply]
Hey @emileond,

Thanks for the notification - generally failure to enable PITR suggests that the Postgres version is not matching. It shouldn't affect MFA.


I'll take a look at the `getAuthenticatorAssuranceLevel` call again on Firefox.

**Comment by fasamoah19** [comment, marked as answer]
I'm running into the same issue. Is there any update on this?

**Reply by tjmills-dev** [reply]
> I'm running into the same issue. Is there any update on this?

See https://github.com/supabase/auth-helpers/pull/455/files - it's patched and if you update to 0.3.0 of the auth helpers it fixes it

**Comment by e2corporation** [comment]
Has this been resolved? This issue is still present in `@supabase/supabase-js@^2.47.10`. I have a verified TOTP device from a custom setup flow, yet the current level is still `aal1` with a `method` of `password` for `currentAuthenticationMethods`.



```jsx
supabase.auth.onAuthStateChange(async (event, session) => {
....
....
if (event === 'SIGNED_IN') {

const { data, error }: AuthMFAGetAuthenticatorAssuranceLevelResponse =
          await supabase.auth.mfa.getAuthenticatorAssuranceLevel()
        const { currentLevel, nextLevel, currentAuthenticationMethods } = data
        
// console.log(currentLevel, currentAuthenticationMethods)
aal1
{
  "method": "password",
  "timestamp": 1735318275
}
```

**Reply by e2corporation** [reply]
Additionally trying to `DELETE` a previously verified TOTP Factor returns this response

```json
{
"code":"insufficient_aal",
"message":"AAL2 required to unenroll verified factor"
}
```

**Reply by kennyward** [reply]
> Additionally trying to `DELETE` a previously verified TOTP Factor returns this response
> 
> ```json
> {
> "code":"insufficient_aal",
> "message":"AAL2 required to unenroll verified factor"
> }
> ```

Did you ever find a fix for this?

Even when i'm at aal2 I get `AAL2 required to unenroll verified factor` error when trying to unenroll a factor. getAuthenticatorAssuranceLevel says i'm at aal2 but the session in the db says aal1, logging out and back in let's me remove the factor, even though getAuthenticatorAssuranceLevel says i'm at aal1 after logging back in.
